torch/fx/examples/vmap.py

```diff
@@ -0,0 +1,177 @@
+# as there is an extra batch dimension in the front. Instead, we must execute
+# `torch.movedim(x,1,3)`. This procedure (and some other stuff to make sure the
+# batch dimension is always at the front) is what's done in
+# `movedim_batching_rule`.
+#
+# There is one final thing to note about these batching rules - they're almost
+# entirely written in normal PyTorch, with the exception of `bdim` attribute
+# that's needed for tracking the batch dimension. That is because in order to
+# use these batching rules, we will be tracing them by passing in `Proxy`
+# objects that will track the operations performed on them and append them to
+# the graph.
+
+def move_bdim_to_front(x, result_ndim=None):
+    x_dim = len(x.shape)
+    x_bdim = x.bdim
+    if x_bdim is None:
+        x = torch.unsqueeze(x, 0)
+    else:
+        x = torch.movedim(x, x_bdim, 0)
+    if result_ndim is None:
+        return x
+    diff = result_ndim - x_dim - (x_bdim is None)
+    for _ in range(diff):
+        x = torch.unsqueeze(x, 1)
+    return x
+
+def movedim_batching_rule(x, from_dim, to_dim):
+    x = move_bdim_to_front(x)
+    return torch.movedim(x, from_dim + 1, to_dim + 1), 0
+
+batching_rules = {}
+def gen_binary_op_batching_rule(op):
+    def binary_op_batching_rule(a, b):
+        a_ndim = len(a.shape)
+        b_ndim = len(b.shape)
+        result_ndim = max(a_ndim, b_ndim)
+        a = move_bdim_to_front(a, result_ndim)
+        b = move_bdim_to_front(b, result_ndim)
+        res = op(a, b)
+        return res, 0
+    return binary_op_batching_rule
+
+def unsqueeze_batching_rule(x, dim):
+    x = move_bdim_to_front(x)
+    if dim >= 0:
+        return torch.unsqueeze(x, dim + 1), 0
+    else:
+        return torch.unsqueeze(x, dim), 0
+
+
+batching_rules[torch.mul] = gen_binary_op_batching_rule(torch.mul)
+batching_rules[torch.unsqueeze] = unsqueeze_batching_rule
+batching_rules[torch.movedim] = movedim_batching_rule
+
+
+# In order to apply a batching rule, we will simply pass in `Proxy` objects as
+# inputs to the functions. As the batching rules need some extra information
+# such as the batch dimension and shape, we will do some bookkeeping here.
+def gen_batching_rule_function(target, *args):
+    def lift_shape(i):
+        res = Proxy(i)
+        res.shape = i.shape
+        res.bdim = i.bdim
+        return res
+    proxy_args = [lift_shape(i) if isinstance(i, fx.Node) else i for i in args]
+    out, bdim = batching_rules[target](*proxy_args)
+    out_node = out.node
+    out_node.bdim = bdim
+    return out_node
+
+def vmap(model: torch.nn.Module, in_axes, example_args) -> torch.nn.Module:
+    """vmap
+    Given a model with inputs, vmap will return a function that works on
+    batched versions of those inputs. Which inputs will be batched is
+    determined by in_axes. In addition, as vmap requires shape (actually
+    rank) information, we will pass in example_args (example inputs for the
+    original module).
```

Comment:
On this function and others please add type annotations to function parameters

Reply:
It's actually quite painful to get it to pass `mypy` for the same reason that you need to add `ignores` to `shape_prop.py`.

Do you think these are necessary for an example? Or do you just want type annotations for documentation purposes.

Reply:
Maybe this pass should have a `Tracer` with `create_node` overridden that creates a subclass of `Node` with the appropriate attributes on it. Going through the docs, I think we should also be encouraging people to use `Tracer.trace` and not `fx.symbolic_trace` generally. (I'm considering deleting `fx.symbolic_trace`)